proposals/reference-types/Overview.md

```diff
@@ -0,0 +1,204 @@
+  - `valtype ::= <numtype> | <reftype>`
+  - locals with reference type are initialised with `null`
+
+* Element types (of tables) are equated with reference types.
+  - `elemtype ::= <reftype>`
+
+* Introduce a simple subtype relation between reference types.
+  - `nullref < t` for all reftypes `t`
+  - `t < anyref` for all reftypes `t`
+  - `anyfunc < anyref`
+
+
+New/extended instructions:
+
+* The new instruction `ref.null` evaluates to the null reference constant.
+  - `ref.null : [] -> [nullref]`
+
+* The new instructions `table.get` and `table.set` access tables.
+  - `table.get $x : [i32] -> [t]` iff `t` is the element type of table `$x`
+  - `table.set $x : [i32 t] -> []` iff `t` is the element type of table `$x`
+  - `table.fill $x : [i32 i32 t] -> []` iff `t` is the element type of table `$x`
+
+* The `call_indirect` instruction takes an additional table index as immediate that identifies the table it calls through.
+  - `call_indirect (type $t) $x : [t1* i32] -> [t2*]` iff `$t` denotes the function type `[t1*] -> [t2*]` and the element type of table `$x` is a subtype of `anyfunc`.
+  - In the binary format, space for the index is already reserved.
+  - For backwards compatibility, the index may be omitted in the text format, in which case it defaults to 0.
+
+
+Table extensions:
+
+* A module may define, import, and export multiple tables.
+  - As usual, the imports come first in the index space.
+  - This is already representable in the binary format.
+
+* Element segments take a table index as immediate that identifies the table they apply to.
+  - In the binary format, space for the index is already reserved.
+  - For backwards compatibility, the index may be omitted in the text format, in which case it defaults to 0.
+
+
+API extensions:
+
+* Any JS object (non-primitive value) or `null` can be passed as `anyref` to a Wasm function, stored in a global, or in a table.
+
+* Any JS function object or `null` can be passed as `anyfunc` to a Wasm function, stored in a global, or in a table.
+
+* Only `null` can be passed as a `nullref` to a Wasm function, stored in a global, or in a table.
+
+
+## Possible Future Extensions
+
+
+### Typed function references
+
+Motivation:
+
+* Allow function pointers to be expressed directly without going through table and dynamic type check.
+* Enable functions to be passed to other modules easily.
+
+Additions:
+
+* Add `(ref $t)` as a reference type
+  - `reftype ::= ... | ref <typeidx>`
+* Add `(ref.func $f)` and `(call_ref)` instructions
+  - `ref.func $f : [] -> (ref $t)  iff $f : $t`
+  - `call_ref : [ts1 (ref $t)] -> [ts2]` iff `$t = [ts1] -> [ts2]`
+* Introduce subtyping `ref <functype> < anyfunc`
+* Subtying between concrete and universal reference types
+  - `ref $t < anyref`
+  - `ref <functype> < anyfunc`
+
+* Typed function references cannot be null!
+
+Question:
+
+* General function have no reasonable default, do we need scoped variables like `let`?
```

Comment:
How would `table.grow`/`Table.prototype.grow` work for a table of function references?  In general, I wonder if having non-null function references is necessary.  E.g., null function references could be implemented without branching as a callable stub (that faults).

Reply:
Good question! I think they both should have an init argument. In the API that defaults to null for backwards compat and convenience.

I added notes to that end.

Reply:
That makes sense.  With the `nullable` type constructor idea, then even host imports or function references could continue to allow a `null` init argument.